### PR TITLE
Basic fusion decision functions of dot operators.

### PR DESCRIPTION
**Pull request by JamesTheZ**
This PR depends on https://github.com/alibaba/BladeDISC/pull/718. It will be rebased after https://github.com/alibaba/BladeDISC/pull/718 is merged.

### CODE REVIEW

**Comment by tanyokwok** [comment]
Nit: Please add a documentation for this computation-intensive codegen module

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 1]
```diff
@@ -1439,6 +1417,8 @@ std::unique_ptr<FusionStrategy> makeNewDeviceStrategy(StringRef device,
     return std::make_unique<StitchGpuFusionStrategy>(options);
   } else if (device == placement_utils::kCpu && strategy == "stitch_base") {
     return std::make_unique<StitchBaseCpuFusionStrategy>(options);
+  } else if (device == placement_utils::kGpu && strategy == "dot") {
+    return std::make_unique<DotGpuFusionStrategy>(options);
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:1421]: Just wonder if there will be a `ConvGpuFusionStrategy` named `conv` in the future ?

File: tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    return initFusionPatternBase(shapeAnalysis, fusion_pattern);
+  }
+
+  // Only one dot.
+  if (dot_ops.size() != 1) {
+    return false;
+  }
+
+  // All ops are supported by CUDA source emitter.
+  for (auto op : mem_intensive_ops) {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc:73]: is this check necessary if `isFusible` works correctly?

**Review by qiuxiafei — COMMENTED** [review]
And, didn't see any invocation on SourceEmitterCuda to generate code. Are they still on the way?

File: tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  // If it is a single supported elementwise op, it may be fused with dot later.
+  // Do not do multi-op fusion for pure element-wise ops to prevent breaking
+  // possible dot fusion.
+  if (dot_ops.size() == 0 && mem_intensive_ops.size() == 1) {
+    return initFusionPatternBase(shapeAnalysis, fusion_pattern);
+  }
+
+  // Only one dot.
```
Comment (linbinskn) [inline tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc:67]: Just curious, there are two constraints about dot in current strategy, first is only on dot and the second is dot is the first op. The first constraint is checked while the second is ignored. Is dot op can always be the first one in current design?

File: tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  fusion_pattern.setFusionType(FusionType::kDot);
+  fusion_pattern.setDominantOp(dot_ops[0]);
+
+  return true;
+}
+
+bool DotGpuFusionStrategy::tryFuse(ShapeAnalysis& shapeAnalysis,
+                                   FusionPattern& lhs, FusionPattern& rhs,
+                                   FusionPattern& target) {
+  auto& op_list = target.getOpList();
```
Comment (linbinskn) [inline tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc:99]: It seems `op_list` is not used.

File: tao_compiler/mlir/disc/disc_compiler.cc [hunk 5]
```diff
@@ -409,6 +408,8 @@ LogicalResult LowerHLOToLLVM(ModuleOp m, const DISCLoweringOptions& options) {
     pm.addNestedPass<FuncOp>(
         disc_ral::createDiscMemRefLoadStoreSimplifierPass());
   }
+  std::string fusion_strategy =
```
[the remark this answers is not in the document]
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/disc_compiler.cc:418]: Fixed.

File: tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc (hunk 2 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc]: It is not necessary. Removed.

File: tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc (hunk 4 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc]: Removed.

**Comment by JamesTheZ** [comment]
> And, didn't see any invocation on SourceEmitterCuda to generate code. Are they still on the way?

This PR is the fusion decision logic, rather than code generation component. The code generation component is https://github.com/alibaba/BladeDISC/pull/720 . The fusion decision logic requires `SourceEmitterCUDA::isSupportedOp` function.

**Pull request merged**